src/node/node_type_deduction.hpp

```diff
@@ -0,0 +1,92 @@
+#pragma once
+#include "napi.h"
+#include "realm/data_type.hpp"
+
+namespace realm {
+namespace js {
+
+struct TypeDeductionImpl {
+    using Value = const Napi::Value;
+    static bool is_bson_type(Value &value, std::string type) {
+        if (!value.IsObject()) {
+            return false;
+        }
+
+        auto object = value.As<Napi::Object>();
+
+        if (object.Has("_bsontype")) {
+            auto bsonType = object.Get("_bsontype");
+            return bsonType.ToString().Utf8Value() == type;
+        }
+
+        return false;
+    }
+    static bool is_decimal128(Value &value) {
+        return TypeDeductionImpl::is_bson_type(value, "Decimal128");
+    }
+
+    static bool is_object_id(Value &value) {
+        return TypeDeductionImpl::is_bson_type(value, "ObjectID");
+    }
+
+    static std::string realm_typeof(DataType value) {
+        std::map<DataType, std::string> realm_typeof = {
+            {DataType::type_String, "string"},
+            {DataType::type_Int, "Int"},
+            {DataType::type_Float, "Float"},
+            {DataType::type_Double, "Double"},
+            {DataType::type_Decimal, "Decimal128"},
+            {DataType::type_Bool, "Boolean"},
+            {DataType::type_ObjectId, "ObjectId"}};
+        return realm_typeof[value];
+    }
+
+    static DataType typeof(Value &value) {
+        if (value.IsNull()) {
```

Comment:
The returned type here and in line 61 is a bit surprising to me. Can you explain why `null` and `undefined` is a `TypedLink`?